library/src/updater.rs

```diff
@@ -805,4 +834,68 @@ mod tests {
         })
         .unwrap();
     }
+
+    #[test]
+    fn no_lock_contention_when_initing_during_patch_check() {
+        let tmp_dir = TempDir::new("example").unwrap();
+        init_for_testing(&tmp_dir, None);
+
+        // Set up the network hooks to sleep for 10 seconds on a patch check request
+        let hooks = NetworkHooks {
+            patch_check_request_fn: |_url, _request| {
+                let patch_check_delay = std::time::Duration::from_secs(10);
+                std::thread::sleep(patch_check_delay);
+                Ok(PatchCheckResponse {
+                    patch_available: false,
+                    patch: None,
+                })
+            },
+            download_file_fn: |_url| Ok([].to_vec()),
+            report_event_fn: |_url, _event| Ok(()),
+        };
+
+        testing_set_network_hooks(
+            hooks.patch_check_request_fn,
+            hooks.download_file_fn,
+            hooks.report_event_fn,
+        );
+
+        let _ = std::thread::spawn(move || {
+            crate::check_for_update().unwrap();
+        });
+
+        let set_config_thread = std::thread::spawn(move || {
+            let tmp_dir = TempDir::new("example").unwrap();
+            for _ in 0..10 {
+                set_config(
+                    crate::AppConfig {
+                        app_storage_dir: tmp_dir.path().display().to_string(),
+                        code_cache_dir: tmp_dir.path().display().to_string(),
+                        release_version: "1.0.0+1".to_string(),
+                        original_libapp_paths: vec!["/dir/lib/arch/libapp.so".to_string()],
+                    },
+                    Box::new(FakeExternalFileProvider {}),
+                    "/dir/lib/arch/libapp.so".into(),
+                    &fake_yaml(),
+                    NetworkHooks {
+                        patch_check_request_fn: |_url, _request| {
+                            Ok(PatchCheckResponse {
+                                patch_available: false,
+                                patch: None,
+                            })
+                        },
+                        download_file_fn: |_url| Ok([].to_vec()),
+                        report_event_fn: |_url, _event| Ok(()),
+                    },
+                );
+            }
+        });
+
+        set_config_thread.join().unwrap();
+
+        // Don't wait for the patch check thread. This test should finish more or less immediately
+        // if set_config isn't waiting for the patch check thread. If set_config is waiting for
+        // the patch check thread, this test will not finish for the duration of patch_check_delay
+        // (defined in patch_check_request_fn above).
```

Comment:
Could probably use a global variable to ensure that we get to the end of the test before the network callback is triggered?

Reply:
Did you one better – there's an `unreachable` in the patch check callback after we sleep, which will cause the test to fail if we're waiting for a response and shouldn't be.